### PR TITLE
Harden gameplay settings JS bridge and add comprehensive tests

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

## Description

What does this PR do? (e.g., "Fixes player jump physics in level 2" or "Adds
new enemy AI script")

## Description

This PR implements a comprehensive suite of unit tests and defensive programming patterns for the `GameplaySettings` module. The primary goal is to resolve **Issue #471** (engine crashes during JavaScript-to-Godot communication) and to ensure the UI menu remains stable during complex lifecycle events and teardowns.

## Key Changes & Bug Fixes

* **Hardened JS Bridge:** Added stricter type, bounds, and payload validation for JavaScript-to-Godot communication. This prevents engine crashes from malformed, empty, or out-of-range `JavaScriptBridge` payloads.
* **Lifecycle & Teardown Safety:** Ensured `GameplaySettings` UI and observers safely handle missing or invalid `Globals` settings, freed nodes, and unexpected tree exits without crashing.
* **UI Synchronization:** Fixed the difficulty label and slider to accurately reflect clamped values from the settings resource, including when values are updated via JavaScript.
* **Signal Safety:** Guarded against duplicate signal connections and unsafe disconnections by checking node validity before wiring or unwiring.
* **Web Overlay State:** Improved teardown logic to consistently clear JS callbacks and web overlay states while seamlessly restoring previous menus when appropriate.
* **Type Flexibility:** Relaxed the `js_window` reference type to a generic `Variant` to better support dictionary-based mocks and flexible JS interfaces.

## Testing

* **Resource Tests:** Added tests for `GameSettingsResource` clamping, signal emission behavior, and boundary values.
* **UI & Reactivity Tests:** Verified that the UI safely synchronizes from the resource, handles slider changes, resets behavior, and propagates observer-driven updates without feedback loops.
* **JS Bridge Tests:** Added coverage for valid payload shapes, malformed inputs, unsupported types, scalar regressions, and missing-node edge cases.
* **Lifecycle Tests:** Verified that cleanup on exit properly disconnects observers, restores previous menus, tears down web overlays, and nullifies callbacks (even if `Globals` is shaky).

## Contributions

* **@ikostan**:
  * Authored the core refactoring logic to resolve **Issue #471**, implementing strict `is_instance_valid` lifecycle guards across the `GameplaySettings` module.
  * Designed and wrote the comprehensive GUT unit test suites covering the resource, UI reactivity, JS bridge, and lifecycle teardown paths.
  * Fixed the critical WebGL/Emscripten crash by correctly implementing standard dot notation (`.length`) for `JavaScriptObject` properties to avoid bridge evaluation errors.
  * Updated and verified E2E Playwright tests (`difficulty_flow_test.py` and `reset_audio_flow_test.py`) to align with the new value-normalization and DOM overlay logic.
  * Integrated and resolved feedback from automated review tools to ensure maximum stability and code quality.

## AI & Bot Contributions

* **@sourcery-ai**:
  * Conducted automated code reviews and identified an edge case in `_on_change_difficulty_js` where `JavaScriptObject` payloads could be improperly rejected, suggesting primitive extraction prior to type-checking.
  * Recommended hardening `_unset_gameplay_settings_window_callbacks` to handle generic `Variant` scenarios safely in non-web or mocked environments.
  * Suggested test suite improvements, including adding `after_each()` teardowns to prevent cross-test pollution and ensuring all JS callbacks (like `_gameplay_reset_cb`) are explicitly asserted as nullified during cleanup tests.
  * Generated detailed PR breakdowns categorizing changes by bug fixes, enhancements, and testing.

* **@coderabbitai**:
  * Formatted and refined the pull request description structure.
  * Generated high-level automated summaries emphasizing defensive signal wiring, missing-node safety, and strict validation of external JS inputs. 
  * Tracked chore-level updates, including the removal of inline citation comments to improve code readability.

* **@deepsource-io**:
  * Conducted automated static analysis and code review.
  * Evaluated the pull request with an overall "A" grade, verifying that the new changes maintain high standards across security, reliability, complexity, and code hygiene.

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [x] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [x] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [x] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [x] Any new unit tests added? (Link to test scene if yes)
- [x] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [x] Code follows Godot style guide (e.g., snake_case for variables)
- [x] No console errors in editor/output
- [x] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Harden GameplaySettings initialization, cleanup, and JS bridge handling while adding comprehensive tests for settings resource, UI synchronization, lifecycle behavior, and JavaScript communication.

Bug Fixes:
- Ensure gameplay settings UI and observers safely handle missing or invalid Globals/settings, freed nodes, and unexpected tree exits without crashing.
- Fix difficulty label and slider to always reflect clamped difficulty values from the settings resource, including values coming from JavaScript.
- Prevent engine crashes and bad state from malformed or scalar JavaScriptBridge payloads by validating types, bounds, and node availability before applying changes.
- Avoid duplicate signal connections and unsafe disconnections by guarding connections and checking node validity before connecting or disconnecting signals.

Enhancements:
- Relax the type of the JS window reference to a generic Variant to support dictionary-based mocks and more flexible JS interfaces.
- Improve gameplay settings teardown to consistently clear JS callbacks and web overlay state while restoring previous menus when appropriate.
- Simplify and clarify Globals settings observer comments without changing behavior.

Tests:
- Add unit tests for GameSettingsResource clamping, signal emission behavior, and default/boundary values.
- Add initialization tests to verify GameplaySettings syncs its UI from the resource, connects observers once, and initializes safely in non-web environments.
- Add UI interaction tests to confirm slider changes, reset behavior, and observer-driven updates propagate correctly without feedback loops.
- Add JavaScriptBridge-focused tests covering valid payload shapes, malformed inputs, unsupported types, scalar regressions, and missing-node safety.
- Add lifecycle tests ensuring cleanup on exit disconnects observers, restores previous menus, tears down web overlays, and nullifies callbacks even with shaky Globals.

## Summary by CodeRabbit

* **Bug Fixes**
  * Hardened gameplay settings UI: defensive signal wiring/teardown, safe handling when UI nodes are missing, and stricter validation/logging for external JS inputs (malformed, empty, out-of-range payloads).
* **Tests**
  * Added comprehensive tests covering resource behavior, UI interactions, lifecycle/teardown, observer reactivity, and JS-bridge edge cases; updated an integration test's expected logs.
* **Chores**
  * Removed inline citation comments for clarity.

### CODE REVIEW

**Comment by ikostan** [comment]
@sourcery-ai review

File: tests/difficulty_flow_test.py [hunk 1]
```diff
@@ -187,12 +189,12 @@ def on_console(msg: Any) -> None:
...
         reset_logs: List[Dict[str, str]] = logs[pre_reset_log_count:]
+
         # Verify that difficulty was reset to the expected default
         assert any(
-            "difficulty" in log["text"].lower()
-            and ("default" in log["text"].lower() or "1.0" in log["text"])
+            "setting 'difficulty' updated to: 1" in log["text"].lower()
             for log in reset_logs
-        ), "Difficulty reset to default was not observed in logs"
+        ), "Resource did not reset difficulty to 1.0 after reset button press"
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/difficulty_flow_test.py]: Thanks, I followed your advice. Please confirm the fix.

File: tests/difficulty_flow_test.py [hunk 2]
```diff
@@ -230,7 +234,8 @@ def on_console(msg: Any) -> None:
         assert page.evaluate("document.getElementById('start-button') !== null")
         page.wait_for_selector("#options-back-button", state="hidden", timeout=2500)
         assert page.evaluate(
-            "document.getElementById('options-back-button') === null || document.getElementById('options-back-button').offsetParent === null"
+            "document.getElementById('options-back-button') === null || document.getElementById("
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/difficulty_flow_test.py]: fixed
